[quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55529 [quant][graphmode][fx] Add shape to nontensor op list
* **#55429 [quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)**

Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan: Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D27609972](https://our.internmc.facebook.com/intern/diff/D27609972)

jerryzh168 committed:
[quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)

Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan: Imported from OSS

Reviewed By: vkuzo

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/utils.py:
```diff
@@ -345,47 +346,58 @@ def create_qparam_nodes(quantizer: QuantizerCls, node_name: str, scale: Any, zer
 
 
-def all_node_args_have_no_tensors(node: Node) -> bool:
+def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module], cache: Optional[Dict[Node, bool]] = None) -> bool:
```
optional: can we make cache a required argument, to reduce complexity?

optional: what broke when trying to implement the cache in a wrapper func, and keep the inner func from knowing about the cache?  I feel like it should be possible, since the cache conceptually is just caching the results of a computation.

jerryzh168 replied:
sure I can make cache required.

for wrapper, I tried that, but since we need to pass around cache to the recursive calls, we ends up with the same code

vkuzo commented on torch/quantization/quantize.py:
```diff
@@ -22,7 +22,8 @@
     return (isinstance(module, torch.quantization.ObserverBase) or
-            isinstance(module, torch.quantization.FakeQuantizeBase))
+            isinstance(module, torch.quantization.FakeQuantizeBase) or
+            isinstance(module, torch.quantization.Logger))  # type: ignore
```
is this the numerical suite logger or something else? what's the context on needing this?

jerryzh168 replied:
this is numeric suite Logger, we need to recognize activation_post_proceess in all_node_args_have_no_tensors after this refactor

jerryzh168 replied:
since we first insert observer and then handle copy node (the handle_copy_nodes function is dealing with a model with observers)

jerryzh168 replied:
btw, sometimes cache doesn't make much sense since the function is called in a local QuantizeHandler

vkuzo replied:
makes sense.  What's the context on having to care about NS loggers here, is there a use case?  It would be great to keep quantization code from not knowing anything about numerical suite, if possible.

jerryzh168 replied:
without this it breaks a test in numeric suite

vkuzo replied:
which test is it?  I don't see it in CI, can you paste a stack trace?

A `_numerical_suite_fx.Logger` is not an observer, I think we should fix the test instead.  Happy to help with that.  I think it's valuable to keep quantization from knowing anything about Numeric Suite directly, due to separation of concerns.

jerryzh168 replied:
Can we do this in a follow up diff? I have other PRs depending on this one that's up for a while. not sure which test is it, but we can see it if we revert this change

jerryzh168 replied:
it's not about the test I think, it's just that our pass needs to know what is the node that's inserted by prepare so that we can skip them in handle_copy_nodes

vkuzo replied:
right, but a Logger is not an observer, so it logically shouldn't be in this function, I don't think we should land this line of code.  If you share the stack trace of the test that's failing, happy to suggest a fix.

vkuzo replied:
if there is some test which is running a quantization pass on a model which has Loggers, then that test needs to be fixed.  I think the right behavior for running a quantization pass on a model with Loggers is "fail with an error message".

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan: Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D27609972](https://our.internmc.facebook.com/intern/diff/D27609972)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan: Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D27609972](https://our.internmc.facebook.com/intern/diff/D27609972)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -327,6 +325,80 @@ def insert_observer_for_input_arg_of_observed_node(
+        root_node, matched_nodes, pattern, quantize_handler, qconfig = matches.get(
+            node.name, (None, None, None, None, None))
+
+        if node.op == "call_module" and is_activation_post_process(modules[node.target]):
```
@vkuzo this is the code that's using the function, it's actually not related to the Logger, we need to use this to handle copy node properly
what `is_activation_post_process` does is to find the nodes that's inserted by prepare. since numeric suite is using prepare to insert Loggers, that's why we need to put Logger in this function

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan: Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D27609972](https://our.internmc.facebook.com/intern/diff/D27609972)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan: Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D27609972](https://our.internmc.facebook.com/intern/diff/D27609972)

[ghstack-poisoned]

vkuzo commented on torch/quantization/__init__.py:
```diff
@@ -44,6 +45,7 @@ def default_eval_fn(model, calib_data):
     # Observers
     'ObserverBase', 'WeightObserver', 'observer', 'default_observer',
     'default_weight_observer', 'default_placeholder_observer',
+    'Logger',
```
should this be removed, since the check was removed?

jerryzh168 replied:
oh, yeah will do. looks like CI is passing

vkuzo commented on torch/quantization/__init__.py:
```diff
@@ -9,6 +9,7 @@
 # from .quantize_fx import *
 from .quantization_mappings import *
 from .fuser_method_mappings import *
+from ._numeric_suite import Logger
```
and this

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function (#54644)"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan: Imported from OSS

Reviewed By: vkuzo

Differential Revision: [D27609972](https://our.internmc.facebook.com/intern/diff/D27609972)

[ghstack-poisoned]
